### ISSUE TITLE
:debug on root of a LiveComponent shows warning

### ISSUE THREAD

**Issue by msaraiva** [msaraiva/surface#142]
Adding a `:debug` directive to the root element of a `LiveComponent` shows the following warning:
```
warning: stateful live components must have a HTML root element
```

Example:

```elixir
  defmodule Counter do
    use Surface.LiveComponent

    def render(assigns) do
      ~H"""
      <div :debug>
        ...
      </div>
      """
    end
  end
```

**Comment by lnr0626** [comment]
The way that debug currently works on html tags when `:code` is specified option (i.e. the default) is essentially by wrapping the node in an `AST.If` with the condition set to true. This was a work around for the limitation created by un-nesting and combining all of the static portions of the template when feeding it into the LiveView EExEngine. While it makes `:debug` work for html nodes, it also introduces this warning (and slightly changes the generated code). Ideally we also would not modify the code to make `:debug` work properly.

A workaround for this warning could be to detect when the root element has `:debug` specified, and special case that; however, I'd really like to find a better way to handle `:debug` for html nodes that doesn't require wrapping the contents with other nodes not specified by the user.

**Comment by msaraiva** [comment]
@lnr0626 would it make sense to have a new `AST.Debug` (or something similar) and use it instead of an `AST.If`? This way the `EExEngine` could print the code as it is without the need to modify it.

**Comment by lnr0626** [comment]
i tried using the `AST.Container` to accomplish something similar - the issue I ran into is that the nodes go through https://github.com/msaraiva/surface/blob/master/lib/surface/compiler/eex_engine.ex#L36-L41 before being printed, which combines the static portions of the html tags.

One possibility would be to check if `:debug` had the appropriate option set in https://github.com/msaraiva/surface/blob/master/lib/surface/compiler/eex_engine.ex#L565 and https://github.com/msaraiva/surface/blob/master/lib/surface/compiler/eex_engine.ex#L549; and if so perform handle that inside those methods. The downside is we'd have to do the transformation on html tags multiple times, but the upside is that it would be accurately printing out the code for that portion of the tree.

As an example, that might look like:

```elixir
  defp to_dynamic_nested_html([
         %AST.Tag{
           element: element,
           attributes: attributes,
           children: children,
           debug: debug_opts
         } = node
         | nodes
       ]) do
    content = [
      "<",
      element,
      to_html_attributes(attributes),
      ">",
      to_dynamic_nested_html(children),
      "</",
      element,
      ">"
    ]

    if Enum.member?(:code, debug_opts) do
      contents
      |> List.flatten()
      |> combine_static_portions()
      |> generate_buffer(...)
      |> maybe_print_expression(node)
    end

    [content, to_dynamic_nested_html(nodes)]
  end
```

**Comment by msaraiva** [comment]
> The downside is we'd have to do the transformation on html tags multiple times

Uhm. That could hurt performance, doesn't it? If it does, unless you can see another option, I think we should just leave it as it is for now. I mean, there's are more important issues to address and since this is a non-blocking-dev-debug-only issue, we can address it some other time. Up to you :)

**Comment by lnr0626** [comment]
yeah - it'd potentially hurt performance pretty badly if a user had a large tree underneath something they marked with `:debug`, or even worse used `:debug` on a large series of nested tags (i.e. `<div :debug><div :debug><div :debug>...</div></div></div>`)

**Comment by msaraiva** [comment]
Well, the performance problem will happen only while debugging (which is kind of common in debugging in general), so I'd say we should do it. Then we can try to improve it in the future in case it becomes a real issue.
